### ISSUE TITLE
Support for passing data based on system parameter "instance"

### ISSUE THREAD

**Issue by concave-sphere** [bevyengine/bevy#2178]
## What problem does this solve or what need does it fill?

I'm working on some code that will involve potentially installing the same system multiple times using parameters that cannot be derived at compile time (i.e, can't be encoded into a type).  In this code, I want to be able to pass data from system X to system Y in an encapsulated way, ideally via a `SystemParam` (chaining is too limiting in this case).

I realize this is vague, but I want to implement something like this (except actually, y'know, useful :grin:)

```rust
struct PerSystemInfo(HashMap<SystemId, String>);

fn build(app: bevy_app::AppBuilder) {
  let per_system_info = HashMap<SystemId, String>;

  let sys = foo.system();
  app.add_system(sys);
  per_system_info.insert(sys.id(), "registration #1")

  // sys2 will have different information than sys
  let sys2 = foo.system();
  app.add_system(sys2);
  per_system_info.insert(sys2.id(), "registration #2");

  app.world_mut().insert_resource(PerSystemInfo(per_system_info));
}

// System param used by the foo function.
//
// It retrieves the private `PerSystemInfo` resource and looks up the
// system info it finds there.  This should happen exactly once, if
// possible (when initializing the app).
struct PerSystemParam(..);

impl PerSystemParam {
  // Returns "registration #1" when running foo from its first registration,
  // and "registration #2" the second time.
  fn per_system_info(&self) -> &'static str { ... }
}
```

## What solution would you like?

Provide a unique identifier for a System.  This exists already (`SystemId`) and it's actually a field of `SystemState`, but there's no accessor for it and I can't find any way to extract it from the black boxes available to `SystemParamState`.  Creating an accessor would allow me to create a resource holding a map of information indexed by `SystemId`, as shown above.

`SystemParamState` should also receive an integer giving the index of the parameter within the parameter list of the system.  This allows disambiguating parameters in the rare case that two parameters have the same type.

So, something like:

```rust
pub unsafe trait SystemParamTrait /* ... */ {
  // ...
  fn init(world: &mut World, system_state: &mut SystemState, param_instance: &SystemParamInstance, config: &Self::Config) -> Self;
}

pub struct SystemParamInstance {
  // Index of this system parameter instance in the function.
  pub arg_index: usize,

  // ID of the system that receives this parameter instance.
  pub system_id: SystemId,
```

### `.with_data`

Another option would be a general syntax for passing parameters to systems, something like:

```
// set up
app.add_system(foo.system().with_data("registration #1"))

// Calling with_data::<T>(T t) on the System causes a Data<T> system
// parameter to be made available to the system.  Like Res, Data<T> is
// an immutable smart pointer to a T.
//
// Any number of Data<T> values can be passed to a given system as long as
// they have distinct types.
//
// e.g, `foo.system().with_data("registration").with_data(1)`.
fn foo(registration: Data<&'static str>) {
  println!("registration");
}
```

Note that in the case above, my `Data` would be a module-private type, and I'd hide it in a custom `SystemParam` that's part of my module API (this is one reason that I can't use `In`: it's a special thing that doesn't work with `SystemParam`).  It's also important that `with_data()` can be invoked multiple times, because the user of my module might want to add their own data items to their system.

This would be more elegant, but probably more complicated to design and implement.  It also can't handle the edge case where two parameters have the same type.

I'd be happy for now with being able to grab the `SystemId` and cobble together my own data pathway using it.

### `SystemParamState::Config`

I wonder if it's possible to use `SystemParamState::Config` for this.  I didn't realize this when first writing this issue, but it appears that it does something like `with_data` above: there's a `config` method on `FunctionSystem` that can mutate the configuration of a system.

I'll take a look and see -- there may _just_ be a way of driving what I need through that loophole.

**Comment by alice-i-cecile** [comment]
Could `Local` resources be extended to support your use case?

**Comment by concave-sphere** [comment]
@alice-i-cecile   The problem is that I specifically want it to _not_ be local ... which I didn't explain very well in my example.

Imagine that the value passed in is a key that describes a location that the system will write.  Later on, another system will get the same key and use it to read the location written by the first system.

(I realize this is all very unclear ... I think I'm close to being able to publish what I'm doing and then I can point at actual code that would be improved)

**Comment by concave-sphere** [comment]
@alice-i-cecile   (earlier comment deleted because it was slightly nonsense)

I think that the `Config` mechanism I described above could be used, with or without `Local` resources.  I'm going to investigate it and see.

**Comment by DJMcNab** [comment]
You seem to be describing `Local<T>` exactly, except with some unneeded complications.

**Comment by concave-sphere** [comment]
Looks like [`config`](https://docs.rs/bevy/0.5.0/bevy/ecs/system/struct.FunctionSystem.html#method.config) does what I need.  I hadn't found it earlier because it's buried on the doc page of a type that is rarely needed -- I stumbled over it while reading Bevy code to see how it works internally.  There may be a way to make it more prominent in the docs -- ideally, the whole DSL of systems would have some top-level documentation, rather than being spread over a bunch of different pages.

I apologize if I wasted everyone's time here.  I know my original explanation of what I'm doing was really inadequate -- `Local` won't work, but it would take a lot of writing to explain why I wanted a different solution to the design goals I'm trying to balance.
